Fix bugs in Leaky ReLU

mejai1206 opened the pull request:
## 🙏 Describe the pull request

- Leaky ReLU의 역전파에서 이전 기울기를 중첩시키지 않는 문제을 수정하고 테스트 케이스를 수정하였습니다.
- Leaky ReLU에서 잘못된 메모리를 접근하여 데이터를 읽는 문제를 수정하고 다양한 크기의 입력으로 테스트 되도록 수정하였습니다.

daemyung commented on tests/test_leaky_relu.py:
```diff
@@ -12,21 +12,53 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import numpy as np
```
`numpy`를 사용하지 말고 `torch`를 사용해주세요. 의존성을 최소화하기 위함입니다.

daemyung commented on tests/test_leaky_relu.py:
```diff
@@ -12,21 +12,53 @@
-    assert util.equal(x, y)
-    assert util.equal(x.grad, y.grad)
+
+def test_module():
```
테스트를 `forward`와 `backward`로 구분하는게 좋을 것 같습니다.

daemyung commented on tests/test_leaky_relu.py:
```diff
@@ -12,21 +12,53 @@
-    assert util.equal(
-        torch.nn.functional.leaky_relu(input2d), trident.function.leaky_relu(input2d)
-    )
+def gen_random_arr2d_list():
```
`@pytest.mark.parametrize` 기능을 사용해서 테스트 케이스에 녹여내는게 좋을 것 같습니다.

mejai1206 replied:
다음과 같이 변경하도록 하겠습니다.

```
rand2d_list1 = [...]
rand2d_list2 = [...]

@pytest.mark.parametrize("input2d_list", [rand2d_list1])
def test_forward(input2d_list):

@pytest.mark.parametrize("input2d_list, target2d_list", [(rand2d_list1, rand2d_list2)])
def test_backward(input2d_list, target2d_list):
```

daemyung replied:
전역변수로 `rand2d_list`가 있는게 보기 좋지 않습니다. 텐서의 크기를 파라미터화시켜서 테스트 케이스 안에서 생성하는게 좋을 것 같습니다.

daemyung commented on tests/test_leaky_relu.py:
```diff
@@ -12,21 +12,53 @@
+        forward_torch = util.train(input_tensor2d_torch, target_tensor2d_torch, torch.nn.LeakyReLU())
+        forward_trident = util.train(input_tensor2d_trident, target_tensor2d_trident, trident.LeakyReLU())
+
+        # check input tensor gradient (TODO: we need to check input's gradient)
```
쓸모 없는 주석은 지워주세요.

daemyung replied:
@mejai1206 주석처리된 코드가 활성화 되야지 테스트가 실패하는거군요? PR에서 테스트만 업데이트 되면 안됩니다. 테스트와 코드가 같이 변경되어야 합니다.

mejai1206 replied:
네 주석 처리된 코드가 Leaky ReLU backward를 검증하는 코드인데, 이 브랜치에선 backward에 문제가 있어서 주석처리를 해뒀습니다. 해당 PR에 backward를 수정한 코드도 함께 반영하고 주석을 풀면 될까요??

daemyung replied:
@mejai1206 최신 코드를 살펴보니 주석이 해제되고 잘못된 구현이 수정된 것 확인하였습니다.

daemyung commented:
@mejai1206 Pull request의 설명은 한글로 작성해도 됩니다.

mejai1206 committed:
Fix bugs in Leaky ReLU

Leaky ReLU backward has two bugs. The first bug is that it doesn't use
gradient outputs. Also test cases don't check the gradient of the input
tensor. The second bug is that it uses invalid pointers to access data.

daemyung commented on tests/test_leaky_relu.py:
```diff
@@ -12,21 +12,43 @@
+def test_backward(shape_list):
+    for shape_0, shape_1 in shape_list:
+        input2d = torch.randn(shape_0, shape_1, device='cuda', requires_grad=False)
+        target2d = torch.randn(shape_0, shape_1, device='cuda', requires_grad=False)
```
변수 이름을 `input`, `target`으로 변경하는게 더 좋을 것 같습니다. 기존에 `input2d`을 사용했던 이유는 `input1d`, `input3d`와 이름 충돌을 방지하기 위해서입니다.

daemyung commented on trident/operation/leaky_relu.py:
```diff
@@ -69,4 +69,4 @@ def get_block_size():
                                         block_size=get_block_size())
 
-        return dx, None
+        return grad_outputs[0] * dx, None
```
`*`보다는 `grad_outputs[0].mm(dx)`가 더 좋을 것 같습니다. [공식 가이드](https://pytorch.org/docs/stable/notes/extending.html#extending-autograd) 참고 해주세요.

mejai1206 replied:
여기선 matrix multiplication이 아닌 elementwise multiplication이 적용 되어야 합니다.
그래서 mm()은 쓸 수 없는데
grad_outputs[0] * dx 와 동일한 elementwise multiplication인
torch.mul(grad_outputs[0], dx) 을 쓰도록 수정하도록 하겠습니다.

daemyung commented on tests/test_leaky_relu.py:
```diff
@@ -12,21 +12,43 @@
+@pytest.mark.parametrize("shape_list",
+                         [[(512, 512), (512, 1024), (1024, 512), (100, 200), (1, 1), (1, 100), (100, 1), (1025, 1)]])
+def test_forward(shape_list):
+    for shape_0, shape_1 in shape_list:
```
`shape_list`를 unpack하지 말고 shape을 사용해서 tensor를 생성하는게 더 깔끔합니다. 아래 예시 참고하세요.

```
shape = (3, 4)
random_tensor = torch.randn(shape)
```

daemyung commented on tests/test_leaky_relu.py:
```diff
@@ -12,21 +12,43 @@
-    assert util.equal(
-        torch.nn.functional.leaky_relu(input2d), trident.function.leaky_relu(input2d)
-    )
+@pytest.mark.parametrize("shape_list",
```
`shape_list`라는 이름보다는 `shapes`가 더 좋은 것 같습니다.

daemyung commented on tests/test_leaky_relu.py:
```diff
@@ -12,21 +12,43 @@
+        input2d_trident = input2d.clone()
+        input2d_trident.requires_grad = True
+
+        target2d_torch = target2d.clone()
```
`target`이 `input`처럼 복사본이 필요한가요? `target`은 `loss`를 계산하기 위해 필요하기 때문에 복사본이 필요하지 않을것 같습니다.

mejai1206 replied:
넵 없어도 될 것 같습니다. 반영하도록 하겠습니다.

daemyung commented on tests/test_leaky_relu.py:
```diff
@@ -12,21 +12,37 @@
-    assert util.equal(
-        torch.nn.functional.leaky_relu(input2d), trident.function.leaky_relu(input2d)
-    )
+@pytest.mark.parametrize("shapes",
```
@mejai1206 마지막으로 Trident에서 문자열은 작은 따옴표를 사용해서 표시하고 있어요. 큰 따옴표를 작은 따옴표로 교체해주세요.

daemyung commented:
@mejai1206 커밋의 제목에는 마침표를 사용하지 않습니다. 그리고 본문의 경우 한 단락으로 작성해주시고 한 줄 당 72자 내로 작성해주세요.

daemyung reviewed (approved):
LGTM
